### PR TITLE
Add v1 Trust Peer Authorization

### PR DESCRIPTION
**Pull request by agunde406**
This PR  includes the message handlers for the new authorization
message, as well as separating out the authorization state machine into
several state machines to make room for new actions/states in the new
trust procedure. This will make it easier in the future to add a
new state machine for challenge authorization.

The v0 Trust is still supported along with v1. If a ConnectRequest
is the first message received v0 trust will be used. If
AuthProtocolRequest is the first message v1 Trust will be used.
This is required to stay backwards compatible with 0.4.

It is also now required to track both our local authorization state
and the state of the remote as separate state machines. This is
because authorization will be happening in parallel.

The new implementation is behind a new experimental features
`trust-authorization`. The `challenge-authorization` features
now depends on this new feature.

### CODE REVIEW

File: libsplinter/src/protocol/authorization.rs [hunk 1]
```diff
@@ -229,8 +229,8 @@ pub struct AuthComplete;
 /// returned.
 #[derive(Debug)]
 pub struct AuthProtocolRequest {
-    auth_protocol_min: i32,
-    auth_protocol_max: i32,
+    pub auth_protocol_min: u32,
+    pub auth_protocol_max: u32,
```
Comment (shannynalayna) [inline libsplinter/src/protocol/authorization.rs:233]: Should these `pub` updates be in the next commit?

Comment (agunde406) [reply libsplinter/src/protocol/authorization.rs:233]: updated the commits to separate adding the pub and the change to u32.

File: libsplinter/Cargo.toml [hunk 2]
```diff
@@ -161,7 +162,7 @@ biome-credentials = ["bcrypt", "biome"]
 biome-key-management = ["biome"]
 biome-notifications = ["biome"]
 biome-profile = ["biome"]
-challenge-authorization = []
+challenge-authorization = ["trust-authorization"]
```
Comment (peterschwarz) [inline libsplinter/Cargo.toml:165]: Why is this dependency needed in this PR?

Comment (agunde406) [reply libsplinter/Cargo.toml:165]: I guess it is not needed yet, but will be as soon as the challenge authorization implementation is added. Ill remove it form this PR.

Comment (agunde406) [reply libsplinter/Cargo.toml]: updated

File: libsplinter/src/network/auth/handlers/v1_handlers.rs [hunk 3]
```diff
@@ -0,0 +1,456 @@
...
+            }
+            Ok(next_state) => panic!("Should not have been able to transition to {}", next_state),
+        }
+        Ok(())
+    }
+}
+
+/// Return the supported protocol version that matches the min/max provided. If there is no
+/// matching protocol version 0 is returned.
+fn supported_protocol_version(min: u32, max: u32) -> u32 {
```
Comment (Caleb-Hill) [inline libsplinter/src/network/auth/handlers/v1_handlers.rs:154]: Should this return a Result instead of an int with a magic value?

Comment (peterschwarz) [reply libsplinter/src/network/auth/handlers/v1_handlers.rs:154]: Probably, but I think this matches how these protocol version matches are checked elsewhere in the code base.  We should add a story to make that sort of conversion, but it shouldn't block this PR

**Review by peterschwarz — APPROVED** [review]

File: libsplinter/src/network/auth/handlers/v1_handlers.rs (hunk 3 above)
Comment (agunde406) [reply libsplinter/src/network/auth/handlers/v1_handlers.rs:154]: I'll add a story.

**Review by Caleb-Hill — APPROVED** [review]

**Review by shannynalayna — APPROVED** [review]

**Pull request merged**